### PR TITLE
Catch and log errors

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes https://github.com/kframework/kore/issues/1132

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Logger/CriticalExecutionError.hs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+newtype CriticalExecutionError =
+    CriticalExecutionError
+    { executionError :: SomeException
+    }
+
+instance Pretty CriticalExecutionError where
+    pretty CriticalExecutionError { executionError } =
+        pretty . displayException $ executionError
+
+instance Entry CriticalExecutionError where
```
Comment (andreiburdusa) [inline kore/src/Kore/Logger/CriticalExecutionError.hs]: Could you explain me why did you add this instance?

Comment (ana-pantilie) [reply kore/src/Kore/Logger/CriticalExecutionError.hs]: @andreiburdusa When we want to log something new, we add a new type which implements the `Entry` class so that we can log it (see https://github.com/kframework/kore/blob/504d615fddb8a3705de36337a78e542a61710a9d/kore/src/Kore/Logger.hs#L243). The reason we make new types for different logging information (instead of just using `LogMessage`, see the functions here https://github.com/kframework/kore/blob/504d615fddb8a3705de36337a78e542a61710a9d/kore/src/Kore/Logger.hs#L148) is that we can use the types themselves to provide additional information in the log and to filter what gets printed. That's my understanding of our logger.

File: kore/app/exec/Main.hs [hunk 2]
```diff
@@ -404,12 +407,13 @@ main = do
...
     let KoreExecOptions { koreLogOptions } = execOptions
-    exitCode <- catch (runLoggerT koreLogOptions go) $
+    exitCode <- runLoggerT koreLogOptions $ catch go $
         \(Goal.WithConfiguration lastConfiguration someException) -> do
-            renderResult
+            liftIO $ renderResult
                 execOptions
                 ("// Last configuration:\n" <> unparse lastConfiguration)
-            throwM someException
+            criticalExecutionError someException
```
Comment (ttuegel) [inline kore/app/exec/Main.hs]: This actually needs to be lifted out to a higher handler because we might have an exception before there is any configuration. The body of the `mainWithOptions` should look something like,

```.hs
    exitCode <- runLoggerT koreLogOptions
        $ handle handleSomeException
        $ handle handleWithConfiguration
        $ go
```
